File: androidDemo/src/main/java/com/novage/p2pml/demo/ui/screens/list/VideoListScreen.kt
package com.novage.p2pml.demo.ui.screens.list

import android.util.Patterns
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.KeyboardArrowRight
import androidx.compose.material.icons.filled.PlayCircleOutline
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.CenterAlignedTopAppBar
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.OutlinedTextFieldDefaults
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.material3.TopAppBarDefaults
import androidx.compose.runtime.Composable
import androidx.compose.runtime.derivedStateOf
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import com.novage.p2pml.demo.data.MediaSample
import com.novage.p2pml.demo.data.VideoStreams

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun VideoListScreen(onVideoSelected: (String, String?) -> Unit) {
    var customUrl by remember { mutableStateOf("") }
    val isUrlValid by remember(customUrl) {
        derivedStateOf {
            if (customUrl.isBlank()) return@derivedStateOf true

            val hasProtocol = customUrl.startsWith("http://") || customUrl.startsWith("https://")
            val isWebUrl = Patterns.WEB_URL.matcher(customUrl).matches()

            hasProtocol && isWebUrl
        }
    }

    val canPlay = remember(customUrl, isUrlValid) {
        customUrl.isNotBlank() && isUrlValid
    }

    Scaffold(
        topBar = {
            Column {
                CenterAlignedTopAppBar(
                    title = {
                        Text(
                            "P2P Media Loader",
                            style = MaterialTheme.typography.titleMedium,
                            fontWeight = FontWeight.SemiBold
                        )
                    },
                    expandedHeight = 44.dp,
                    colors = TopAppBarDefaults.topAppBarColors(
                        containerColor = MaterialTheme.colorScheme.background,
                        titleContentColor = MaterialTheme.colorScheme.onBackground
                    )
                )
                HorizontalDivider(color = MaterialTheme.colorScheme.outlineVariant)
            }
        },
        containerColor = MaterialTheme.colorScheme.background
    ) { innerPadding ->
        LazyColumn(
            modifier = Modifier
                .fillMaxSize()
                .padding(innerPadding)
                .padding(horizontal = 16.dp),
            verticalArrangement = Arrangement.spacedBy(8.dp),
            contentPadding = PaddingValues(bottom = 24.dp)
        ) {
            item {
                Column(
                    verticalArrangement = Arrangement.spacedBy(8.dp),
                    modifier = Modifier.padding(top = 16.dp)
                ) {
                    OutlinedTextField(
                        value = customUrl,
                        onValueChange = { customUrl = it },
                        placeholder = {
                            Text(
                                "Paste Your Manifest URL",
                                color = MaterialTheme.colorScheme.onSurfaceVariant
                            )
                        },
                        modifier = Modifier.fillMaxWidth(),
                        singleLine = true,
                        shape = RoundedCornerShape(8.dp),
                        isError = !isUrlValid && customUrl.isNotEmpty(),
                        colors = OutlinedTextFieldDefaults.colors(
                            focusedContainerColor = MaterialTheme.colorScheme.surface,
                            unfocusedContainerColor = MaterialTheme.colorScheme.surface,
                            errorContainerColor = MaterialTheme.colorScheme.surface,
                            unfocusedBorderColor = Color.Transparent,
                            focusedBorderColor = Color.Transparent,
                            errorBorderColor = MaterialTheme.colorScheme.error,
                            cursorColor = MaterialTheme.colorScheme.primary
                        )
                    )

                    if (!isUrlValid && customUrl.isNotEmpty()) {
                        Text(
                            text = "Enter a valid URL starting with http:// or https://",
                            color = MaterialTheme.colorScheme.error,
                            style = MaterialTheme.typography.bodySmall
                        )
                    }

                    Button(
                        onClick = { onVideoSelected(customUrl.trim(), null) },
                        enabled = canPlay,
                        shape = RoundedCornerShape(8.dp),
                        contentPadding = PaddingValues(horizontal = 24.dp, vertical = 16.dp),
                        colors = ButtonDefaults.buttonColors(
                            containerColor = MaterialTheme.colorScheme.primary,
                            disabledContainerColor = MaterialTheme.colorScheme.onSurface.copy(alpha = 0.12f)
                        ),
                        modifier = Modifier.fillMaxWidth()
                    ) {
                        Text("Play URL", fontWeight = FontWeight.Bold)
                    }
                }
            }

            item {
                HorizontalDivider(
                    color = MaterialTheme.colorScheme.outlineVariant,
                    modifier = Modifier.padding(vertical = 8.dp)
                )
            }

            item {
                Text(
                    text = "Samples",
                    style = MaterialTheme.typography.titleMedium,
                    color = MaterialTheme.colorScheme.onSurfaceVariant
                )
            }

            items(VideoStreams.samples) { stream ->
                StreamListItem(stream) {
                    onVideoSelected(stream.uri, stream.customEngineUrl)
                }
            }
        }
    }
}

@Composable
private fun StreamListItem(stream: MediaSample, onClick: () -> Unit) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .clickable { onClick() }
            .background(MaterialTheme.colorScheme.surface, shape = RoundedCornerShape(8.dp))
            .padding(16.dp),
        verticalAlignment = Alignment.CenterVertically
    ) {
        Icon(
            imageVector = Icons.Default.PlayCircleOutline,
            contentDescription = null,
            tint = MaterialTheme.colorScheme.primary
        )

        Spacer(modifier = Modifier.width(16.dp))

        Column(modifier = Modifier.weight(1f)) {
            Text(stream.title, fontWeight = FontWeight.Bold)
            Text(
                stream.description,
                maxLines = 1,
                overflow = TextOverflow.Ellipsis,
                style = MaterialTheme.typography.bodySmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant
            )
        }

        Spacer(modifier = Modifier.width(16.dp))

        Icon(
            imageVector = Icons.AutoMirrored.Filled.KeyboardArrowRight,
            contentDescription = null,
            tint = MaterialTheme.colorScheme.onSurfaceVariant
        )
    }
}

File: iosDemo/iosDemo/UI/Screens/List/VideoListScreen.swift
import SwiftUI

struct VideoListScreen: View {
    @State private var customUrl: String = ""

    var isUrlValid: Bool {
        guard !customUrl.isEmpty else { return true }
        guard let url = URL(string: customUrl),
              let scheme = url.scheme,
              scheme == "http" || scheme == "https",
              url.host != nil else { return false }
        return true
    }

    var canPlay: Bool {
        !customUrl.isEmpty && isUrlValid
    }

    var body: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: 16) {
                VStack(alignment: .leading, spacing: 8) {
                    TextField("Paste Your Manifest URL", text: $customUrl)
                        .padding()
                        .background(AppTheme.surface)
                        .cornerRadius(8)
                        .foregroundColor(AppTheme.onSurface)
                        .keyboardType(.URL)
                        .textInputAutocapitalization(.never)
                        .overlay(
                            RoundedRectangle(cornerRadius: 8)
                                .stroke(!isUrlValid && !customUrl.isEmpty ? AppTheme.error : Color.clear, lineWidth: 1)
                        )

                    if !isUrlValid, !customUrl.isEmpty {
                        Text("Enter a valid URL starting with http:// or https://")
                            .font(.caption)
                            .foregroundColor(AppTheme.error)
                    }

                    NavigationLink(value: AppRoute.player(
                        videoUrl: customUrl.trimmingCharacters(in: .whitespacesAndNewlines),
                        customEngineUrl: nil
                    )) {
                        Text("Play URL")
                            .fontWeight(.bold)
                            .frame(maxWidth: .infinity)
                            .padding()
                            .background(canPlay ? AppTheme.primary : AppTheme.onSurface.opacity(0.12))
                            .foregroundColor(canPlay ? AppTheme.onPrimary : AppTheme.onSurface.opacity(0.38))
                            .cornerRadius(8)
                    }
                    .disabled(!canPlay)
                }
                .padding(.top, 16)

                Divider().background(AppTheme.onSurfaceVariant)

                Text("Samples")
                    .font(.title3)
                    .fontWeight(.medium)
                    .foregroundColor(AppTheme.onSurfaceVariant)

                LazyVStack(spacing: 8) {
                    ForEach(VideoStreams.samples, id: \.self) { stream in
                        NavigationLink(value: AppRoute.player(
                            videoUrl: stream.uri,
                            customEngineUrl: stream.customEngineUrl
                        )) {
                            HStack(spacing: 16) {
                                Image(systemName: "play.circle")
                                    .foregroundColor(AppTheme.primary)
                                    .font(.title2)

                                VStack(alignment: .leading, spacing: 4) {
                                    Text(stream.title)
                                        .fontWeight(.bold)
                                        .foregroundColor(AppTheme.onBackground)
                                        .lineLimit(nil)
                                        .fixedSize(horizontal: false, vertical: true)
                                        .multilineTextAlignment(.leading)
                                    Text(stream.description)
                                        .font(.caption)
                                        .foregroundColor(AppTheme.onSurfaceVariant)
                                        .lineLimit(1)
                                }
                                Spacer()
                                Image(systemName: "chevron.right")
                                    .foregroundColor(AppTheme.onSurfaceVariant)
                            }
                            .padding()
                            .background(AppTheme.surface)
                            .cornerRadius(8)
                        }
                    }
                }
                .padding(.bottom, 24)
            }
            .padding(.horizontal, 16)
        }
        .background(AppTheme.background.edgesIgnoringSafeArea(.all))
        .navigationTitle("P2P Media Loader")
        .navigationBarTitleDisplayMode(.inline)
    }
}
